### PR TITLE
Improve detector 2theta coverage calculation in SofQWNormalisedPolygon

### LINKED ISSUE #24305: SofQWNormalisedPolygons: detector angular widths not optimal
`SofQWNormalisedPolygons` computes the angular (2theta) width of a detector in the 'up' direction only. This is not accurate, as the detectors can be wider than tall, or, depending on the detector position, an angular width taken in the horizontal plane can be larger. To get more accurate results, the algorithm could calculate the angular widths in all three directions, 'up', 'beam' and 'in plane' and choose the greatest one.

To have really accurate angular width, the algorithm could accept a table of per-detector angular widths, precalculated by whatever method is deemed best. As most instruments are static in configuration, this would avoid excessive repetitive computations in the algorithm itself.

### PR DESCRIPTION
**Pull request by soininen**
The method of calculating detector's 2theta coverage by picking a point in the bounding box's center and top is not very accurate. This PR improves the calculation in several ways:
* Instead of expressing the 2theta coverage as nominal 2theta +/- width/2, we now use just the minimum and maximum 2theta. The nominal 2theta does not always correspond to the actual average 2theta.
* Check if a detector has a simple shape where the 2thetas can be calculated accurately. In case of cuboid and cylinder, we do just this.
* More complicated detector shapes will still be handled through the bounding box. However, we now pick points on each side of the box and compute the 2thetas for those.
* There is now an additional property `DetectorTwoThetaRanges` which accepts a `TableWorkspace` tabulating the per-detector min and max 2thetas. Handy for overriding the calculation in the algorithm.

**Report to:** [nobody].

**To test:**

With unit test data in Mantid's data search directories, the following script should run without problems:
```python
ws = DirectILLCollectData('ILL/IN4/084447.nxs')
ws = ConvertUnits(ws, 'DeltaE', 'Direct')
ws = Rebin(ws, [0.02])
sqw = SofQWNormalisedPolygon(ws, [0., 0.04, 4.], 'Direct')
```

Observe the gap (NaNs) between the low angle (small Q, around zero energy transfer) and large angle detectors.

To test the new 2theta table input property, this file can be used:
[2theta_table.zip](https://github.com/mantidproject/mantid/files/2732471/2theta_table.zip)
The script changes to: 
```python
ws = DirectILLCollectData('ILL/IN4/084447.nxs')
ws = ConvertUnits(ws, 'DeltaE', 'Direct')
ws = Rebin(ws, [0.02])
sqw = SofQWNormalisedPolygon(ws, [0., 0.04, 4.], 'Direct', DetectorTwoThetaRanges='2theta')
```

The 2thetas in the above file are calculated using an analytical formula from the instrument scientists --- the results should be close to identical when not using the pre-calcualted table *except* for the small angle detectors (low Q) for which the analytical formula is wrong.

Fixes #24305.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/SofQWNormalisedPolygon.cpp [hunk 1]
```diff
@@ -21,38 +24,259 @@
 #include "MantidKernel/PhysicalConstants.h"
 #include "MantidTypes/SpectrumDefinition.h"
 
+#include <boost/math/special_functions/pow.hpp>
+
 using Mantid::Geometry::rad2deg;
+using boost::math::pow;
```
Comment (mganeva) [inline Framework/Algorithms/src/SofQWNormalisedPolygon.cpp]: Do you have any specific reason for `boost::math::pow` and not `std::pow`?

File: Framework/Algorithms/src/SofQWNormalisedPolygon.cpp [hunk 2]
```diff
@@ -21,38 +24,259 @@
...
 
 namespace {
 /**
- * @brief Calculate the min and max 2theta of a detector.
- * The 2theta are calculated only for the centre point and the 'top'
- * point, assuming that the width is zero. This works adequately well
- * for high aspect ratio detectors.
+ * Calculate 2theta for a point in detector's local coordinates.
+ * @param detInfo a detector info
+ * @param detIndex an index to the detector info, not detector ID
```
Comment (mganeva) [inline Framework/Algorithms/src/SofQWNormalisedPolygon.cpp]: wouldn't it better to name it `detInfoIndex`?

File: Framework/Algorithms/src/SofQWNormalisedPolygon.cpp [hunk 3]
```diff
@@ -177,8 +400,8 @@ void SofQWNormalisedPolygon::exec() {
       const V2D ur(dE_jp1, m_EmodeProperties.q(dE_jp1, thetaUpper, det));
       const V2D ul(dE_j, m_EmodeProperties.q(dE_j, thetaUpper, det));
       if (g_log.is(Logger::Priority::PRIO_DEBUG)) {
-        logStream << "Spectrum=" << specNo << ", theta=" << theta
-                  << ",thetaWidth=" << thetaWidth << ". QE polygon: ll=" << ll
+        logStream << "Spectrum=" << specNo << ", lower theta=" << thetaLower
```
Comment (mganeva) [inline Framework/Algorithms/src/SofQWNormalisedPolygon.cpp]: Is there any particular reason to output these angles in radians? If not, I'd recommend to output them in degrees to be consistent.

File: Framework/Geometry/inc/MantidGeometry/Instrument/ComponentHelper.h [hunk 4]
```diff
@@ -25,8 +25,7 @@ class ParameterMap;
 
 /**
   A set of helper functions for dealing with components, i.e. movement, rotation
-  that require
-  interaction with the ParamterMap.
+  that require interaction with the ParamterMap.
```
Comment (mganeva) [inline Framework/Geometry/inc/MantidGeometry/Instrument/ComponentHelper.h]: Typo: I think it should be `ParameterMap`

**Review by mganeva — COMMENTED** [review]
Here a couple of comments on the code

**Review by mganeva — CHANGES_REQUESTED** [review]
Please, mention in `DetectorTwoThetaRanges` property description (`SofQW.cpp`, line 111) that this property is used only by `SofQWNormalizedPolygon`. Otherwise users of other algorithms might get confused.
See also previous comments on the code.

File: Framework/Algorithms/src/SofQWNormalisedPolygon.cpp (hunk 1 above)
Comment (soininen) [reply Framework/Algorithms/src/SofQWNormalisedPolygon.cpp]: Mostly out of an old habit as both produce the same code when optimizations are turned on. However, in debug mode `boost::pow<>()` produces nicer assembly.

**Review by mganeva — APPROVED** [review]

**Comment by mganeva** [comment]
@soininen There is a typo in the doc string in line 115 in SofQW.cpp (`use` -> `used`). Just fix it in your next PR.

The changes make the code look better. The algorithm works as described. Unit tests and documentation are provided. Detector coverage is calculated more accurate now. Tested for IN4, TOFTOF and DNS TOF data.

File: Framework/Algorithms/src/SofQWNormalisedPolygon.cpp [hunk 5]
```diff
@@ -165,20 +379,22 @@ void SofQWNormalisedPolygon::exec() {
...
-                  << ", lr=" << lr << ", ur=" << ur << ", ul=" << ul << "\n";
+        logStream << "Spectrum=" << specNo
+                  << ", lower theta=" << thetaLower * rad2deg
+                  << ", upper theta=" << thetaUpper * rad2deg
+                  << ". QE polygon: ll=" << ll << ", lr=" << lr << ", ur=" << ur
+                  << ", ul=" << ul << "\n";
       }
 
       using FractionalRebinning::rebinToFractionalOutput;
-      rebinToFractionalOutput(Quadrilateral(std::move(ll), std::move(lr),
```
Comment (martyngigg) [inline Framework/Algorithms/src/SofQWNormalisedPolygon.cpp]: I assume the moves of the V2D made very little difference to the performance?

Comment (soininen) [reply Framework/Algorithms/src/SofQWNormalisedPolygon.cpp]: I might be mistaken but I don't think any moves took place here.
1. `Quadrilateral`'s constructror takes its arguments by normal reference
2. `ll`, `lr`, `ur` and `ul` are marked as `const` above
3. What is there to move in `V2D`?

Comment (martyngigg) [reply Framework/Algorithms/src/SofQWNormalisedPolygon.cpp]: Ah I assumed `Quadrilateral` took the arguments by value. Never mind :)

**Pull request merged**